db/migrate/20200331111947_add_reporting_org_association_to_activity.rb

```diff
@@ -0,0 +1,37 @@
+class AddReportingOrgAssociationToActivity < ActiveRecord::Migration[6.0]
+  def up
+    add_reference :activities, :reporting_organisation, type: :uuid, foreign_key: {to_table: :organisations}
+
+    ActiveRecord::Base.transaction do
+      activities = Activity.where(reporting_organisation: nil)
+      service_owner = Organisation.find_by_service_owner(true)
+
+      activities.each do |activity|
+        organisation = activity.organisation
+        activity.reporting_organisation = if organisation.is_government?
+          service_owner
+        else
+          organisation
+        end
+        activity.save(validate: false)
```

Comment:
I _think_ that with `save` (without a bang) and `validate: false` being used here the value a transaction block provides is reduced. It might only be helpful now if there are connection problems with the database?

It would be great if we could assume that when working with our records that they are always in a valid state and change this to: `activity.save` 

This would help ensure that the records stay valid and help us prevent introduced new problems unwittingly by using `validate: false`. Is there a case or database state you've seen that has made you cautious? Maybe we can restore that integrity first separately so that this can run with confidence. If we went to each environment and asked all activities if they were `valid?` would that give us that confidence?

Reply:
I used `validate: false` because we save Activities to the database which are incomplete. So if, for example, we were updating an Activity which hadn't get been completed (e.g. if the user hadn't yet added a geography or whatever) then the Activity reporting organisation would still be saved even if it's missing a geography. 

If you don't think there will be any incomplete activities in prod then I suppose we can leave it out, but I worry if there are any, we will end up with activities without reporting orgs (because there is no way to add a reporting org via the UI)

Reply:
Ah I see what you mean. Hopefully we'll still be able to use the `wizard_status` in the conditional validations to allow us to update those activities?

Reply:
I'm not sure... I added the `validate: false` because I had this problem running the migration locally! Some Activities wouldn't update because they were missing attributes.

Reply:
I've replicated data from production to check what this is. You're right it is missing `geography`, we must have forgot to migrate the data when adding that field. I will create a migration that if merged first will allow us to use `save!` here.

There was a further problem with a hole in our conditional validation. The identifier field is required and must be unique, these are both being run with the first wizard_status: :identifier. I think we can fix this conditional validation as a separate problem by adding a new "blank" step. This will allow us to call `save!` here!

Reply:
1. https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/pull/310

Reply:
1. https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/pull/311